Gems/ROS2/Code/Source/Odometry/ROS2OdometryCovariance.cpp

```diff
@@ -0,0 +1,58 @@
+/*
+ * Copyright (c) Contributors to the Open 3D Engine Project.
+ * For complete copyright and license terms please see the LICENSE at the root of this distribution.
+ *
+ * SPDX-License-Identifier: Apache-2.0 OR MIT
+ *
+ */
+
+#include "ROS2OdometryCovariance.h"
+#include "AzCore/RTTI/RTTIMacros.h"
+#include "AzCore/base.h"
+#include <AzCore/Serialization/EditContext.h>
+#include <AzCore/Serialization/EditContextConstants.inl>
+#include <array>
+#include <cstddef>
+
+namespace ROS2
+{
+    void ROS2OdometryCovariance::Reflect(AZ::ReflectContext* context)
+    {
+        if (AZ::SerializeContext* serialize = azrtti_cast<AZ::SerializeContext*>(context))
+        {
+            serialize->Class<ROS2OdometryCovariance>()
+                ->Version(1)
+                ->Field("Linear covariance", &ROS2OdometryCovariance::m_linearCovariance)
+                ->Field("Angular covariance", &ROS2OdometryCovariance::m_angularCovariance);
+
+            if (AZ::EditContext* ec = serialize->GetEditContext())
+            {
+                ec->Class<ROS2OdometryCovariance>("ROS2 Odometry Covariance", "Define Odometry covariance")
+                    ->DataElement(
+                        AZ::Edit::UIHandlers::Default,
+                        &ROS2OdometryCovariance::m_linearCovariance,
+                        "Linear covariance",
+                        "Set ros linear covariance")
+                    ->DataElement(
+                        AZ::Edit::UIHandlers::Default,
+                        &ROS2OdometryCovariance::m_angularCovariance,
+                        "Angular covariance",
+                        "Set ros angular covariance");
+            }
+        }
+    }
+
+    std::array<double, 36> ROS2OdometryCovariance::GetRosCovariance() const
+    {
+        constexpr size_t width = 6;
+        std::array<double, 36> ros2Covariance;
+        ros2Covariance[width * 0 + 0] = m_linearCovariance.GetX();
+        ros2Covariance[width * 1 + 1] = m_linearCovariance.GetY();
+        ros2Covariance[width * 2 + 2] = m_linearCovariance.GetZ();
+        ros2Covariance[width * 3 + 3] = m_angularCovariance.GetX();
+        ros2Covariance[width * 4 + 4] = m_angularCovariance.GetY();
+        ros2Covariance[width * 5 + 5] = m_angularCovariance.GetZ();
+        return ros2Covariance;
```

Comment:
suggestion: We can make this more readable and self explanatory if we laid out the vector line up the values.

```suggestion
        const AZ::Vector3 & lin = m_linearCovariance;
        const AZ:Vector3 & ang = m_angularCovariance;
        
        return {
            lin.GetX(),  0.0d,       0.0d,       0.0d,       0.0d,       0.0d, 
            0.0d,        lin.GetY(), 0.0d,       0.0d,       0.0d,       0.0d,
            0.0d,        0.0d,       lin.GetZ(), 0.0d,       0.0d,       0.0d,
            0.0d,        0.0d,       0.0d,       ang.GetX(), 0.0d,       0.0d,
            0.0d,        0.0d,       0.0d,       0.0d,       ang.GetY(), 0.0d, 
            0.0d,        0.0d,       0.0d,       0.0d,       0.0d,       ang.GetZ() 
        };

```